modulemd/modulemd-module-stream.c

```diff
@@ -942,12 +950,75 @@ stream_expansion_helper (ModulemdDependencies *deps,
 
       /*
        * If a module is present in the dependency list but has no associated
-       * streams (which corresponds to "modulename: []" in the spec), the
-       * intention is to expand the list to be all active existing streams for
-       * the module. Unfortunately, that is something only the Module Build
-       * Service can do, so we must fail the stream expansion.
+       * streams (which corresponds to "modulename: []" in the spec), or it
+       * contains only negated items (corresponding to
+       * "modulename: [-stream]"), the intention is to expand the list to be
+       * all active existing streams for the module. Check to see if we have a
+       * listing of known streams to use. We can assume if the first item
+       * begins with a '-' that they all do, since that is checked during the
+       * parsing and validation of the YAML.
        */
-      if (g_strv_length (streams) == 0)
+      if (g_strv_length (streams) == 0 || streams[0][0] == '-')
+        {
+          g_debug ("Ambiguous expansion detected for %s", module);
+          if (helper)
+            {
+              known_streams =
+                modulemd_upgrade_helper_get_known_streams_as_array (helper,
+                                                                    module);
+
+              /* Check whether we got back at least one stream */
+              if (known_streams->len > 0)
+                {
+                  /* Iterate through the negations */
+                  for (j = 0; NULL != streams[j]; j++)
+                    {
+                      /* Skip the leading '-' */
+                      stream = streams[j] + 1;
+
+                      /* Remove it if it exists in the list */
+                      found = g_ptr_array_find_with_equal_func (
+                        known_streams, stream, g_str_equal, &idx);
+                      if (found)
+                        {
+                          g_ptr_array_remove_index_fast (known_streams, idx);
```

Comment:
```suggestion
                          g_ptr_array_remove_index (known_streams, idx);
```
Funny this popped up immediately after our discussion about ordering. Could you switch this to the slower order-preserving variation until I figure out how to sort lists of `ModulemdBuildConfig`s?

Reply:
Heh, okay. I made the requested changes.